### PR TITLE
Kickstart FSDP2

### PR DESCRIPTION
**Pull request by bowenyang008**
# What does this PR do?

This PR kicks off the support of Pytorch's FSDP2. It lays down the initial setup, including the bare minimum src implementation, an initial unit test for base functionality (and weight tying) and a doc file.

The general approach is to start supporting the default/base functionality of FSDP2/fully_shard API and incrementally migrate from FSDP1 or build additional functionalities on top of it according to the fsdp2.md file in following PRs.

# What issue(s) does this change relate to?

FSDP2: motivation and plan can be found in the fsdp2.md file and [here](https://docs.google.com/document/d/1M__9POM1XPFr4RXsRJ7UceZqaFC764EmH8fCbmauZCU/edit?tab=t.0#heading=h.4cz970y1mk93). 

# Testing

One unit test to test sub helper functions and one new unit test that mimics FSDP1 test to test weight tying correctness with Trainer. Accuracy (overfit to dataset) is manually checked, proper Acc test would be added later.

### CODE REVIEW

File: composer/distributed/fsdp2.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        mp_policy (Optional[MixedPrecisionPolicy]): Mixed precision policy. Default: None
+        offload_policy (Optional[OffloadPolicy]): Offloading policy. Default: None
+    """
+    device_mesh: Optional[DeviceMesh] = None
+    reshard_after_forward: Union[bool, int] = True
+    mp_policy: Optional[MixedPrecisionPolicy] = None
+    offload_policy: Optional[OffloadPolicy] = None
+
+
+def identify_shardable_modules(modules: list[nn.Module]) -> tuple[list[nn.Module], set[nn.Module]]:
```
Comment (dakinggg) [inline composer/distributed/fsdp2.py:37]: would be good to add some unit tests for each of these helper functions

File: composer/core/state.py [hunk 2]
```diff
@@ -877,8 +877,14 @@ def evaluators(self, evaluators: Union[Evaluator, Sequence[Evaluator]]):
     def fsdp_enabled(self):
         """Indicates if FSDP is enabled."""
         for module in self.model.modules():
+            # FSDP is FSDP1, FSDPModule is FSDP2
             if isinstance(module, FSDP):
                 return True
+            # TODO(boweny) remove this once we deprecate torch-cpu 2.5
```
[the remark this answers is not in the document]
Comment (bowenyang008) [reply composer/core/state.py:883]: does it make sense to create a JIRA ticket and have TODO(ticket) to record it. If I remove this TODO and just doc it in planning, it would be hard to track where the relevant changes are.

File: composer/distributed/fsdp2.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bowenyang008) [reply composer/distributed/fsdp2.py:37]: yup, basically filtering modules to two categories: w/ (so in its own FSDP block) and w/o (so in the root model's fsdp block) shared params

Comment (bowenyang008) [reply composer/distributed/fsdp2.py:37]: sg, will do

File: composer/distributed/fsdp2.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+from typing import Optional, Union
+
+from torch import nn
+from torch.distributed._tensor.device_mesh import DeviceMesh
+from torch.distributed.fsdp._fully_shard import fully_shard
+from torch.distributed.fsdp._fully_shard._fsdp_api import MixedPrecisionPolicy, OffloadPolicy
+
+
+@dataclass
+class FSDP2Config:
```
[the remark this answers is not in the document]
Comment (bowenyang008) [reply composer/distributed/fsdp2.py:17]: Meta init actually works. e.g., in the unit test I manually called param init for meta device. it is just not integrated into the API yet

**Review by dakinggg — APPROVED** [review]

File: tests/trainer/test_fsdp2.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    PartialWeightTiedModel,
+    RandomClassificationDataset,
+    SimpleWeightTiedModel,
+    world_size,
+)
+
+_INIT_DEVICES = ['cuda', 'meta']
+
+
+@pytest.mark.parametrize('model', [SimpleWeightTiedModel, PartialWeightTiedModel])
```
[the remark this answers is not in the document]
Comment (bowenyang008) [reply tests/trainer/test_fsdp2.py:29]: fixed
